test/cri-containerd/main_test.go

```diff
@@ -159,6 +159,13 @@ func requireBinary(t *testing.T, binary string) string {
 }
 
 func getWindowsNanoserverImage(build uint16) string {
+	// Due to some efforts in improving down-level compatibility for Windows containers (see
+	// https://techcommunity.microsoft.com/t5/containers/windows-server-2022-and-beyond-for-containers/ba-p/2712487)
+	// the ltsc2022 image should continue to work on builds ws2022 and onwards. The panic for "unsupported build"
+	// should only be triggered if the user is on 21h1 or a build older than RS5 now.
+	if build > osversion.V21H2Server {
```

Comment:
For any release that is > V21H2Server we will always end up using the "nanoserver:ltsc2022" image. Right now this is fine because we don't have any such release but in the future this could cause problems. How about moving this if condition under the `default` case? In `default` case we check if build > osversion.V21H2Server and use the nanoserver:ltsc2022 image, otherwise we panic.

Reply:
Sure that works for me

Reply:
Done @ambarve

Reply:
I think I liked the previous way better. The future evolution in my mind was adding new if checks before that `build > osversion`, so the "default" clause didn't grow too long and nested, since over time this will grow as well, as we eventually hit OS build versions that can't run the ltsc2022 image, whenever that happens. It also loses the separation of "host build ➡ image build" mapping from the "image build ➡ MCR repository string", which I liked about that layout.

Structuring it the way it is after 6388b446cf5cef3f4e951d6ad5c865dcba364c36, I think it'd be clearer to go with the `getWindowsServercoreImage` approach before af9e01f92621cccf296c7c57130bc1a1f8e3a9fe instead, because that's what we've kind-of done, except using `default` to add an extra level of nesting for the non `==` cases.

Reply:
I am not sure I understand your point about growing the "default" clause. As new OS builds and corresponding images are released, we will add a new "case" for those in the switch statement and if your machine doesn't match any of those we get into the default case which can remain as it is for at least next few years.

Reply:
It's possible I've misplaced some expectations. I had understood that there were going to be plenty of releases, e.g. Windows 11, Azure Stack HCI, where the new build would _not_ have a corresponding container base image, so over time the code would grow to look like:
```
	switch build {
	case osversion.RS5:
		return "mcr.microsoft.com/windows/nanoserver:1809"
	case osversion.V19H1:
		return "mcr.microsoft.com/windows/nanoserver:1903"
	case osversion.V19H2:
		return "mcr.microsoft.com/windows/nanoserver:1909"
	case osversion.V20H1:
		return "mcr.microsoft.com/windows/nanoserver:2004"
	case osversion.V20H2:
		return "mcr.microsoft.com/windows/nanoserver:2009"
	default:
                if build >= osversion.V27H2Server {
			return "mcr.microsoft.com/windows/nanoserver:ltsc2038"
                } else if build >= osversion.V24H2Server {
			return "mcr.microsoft.com/windows/nanoserver:ltsc2025"
		} else if build >= osversion.V21H2Server {
			return "mcr.microsoft.com/windows/nanoserver:ltsc2022"
		}
		panic("unsupported build")
	}
}
```
i.e. the only thing that would grow over time would be the `if` clause in the default block in the current layout.

Reply:
To illustrate what I'm thinking: I was doing this work, I'd have refactored the behaviour along the lines of the mapping I described earlier:
```go
// Manage cross-compatible container/host build rules
// https://techcommunity.microsoft.com/t5/containers/windows-server-2022-and-beyond-for-containers/ba-p/2712487
def getContainerBuildForOSBuild(build uint16) uint16 {
    if build >= osversion.V21H2Server {
        return osversion.V21H2Server
    }
    return build
}

def getContainerTagForContainerBuild(build uint16) string {
	switch build {
	case osversion.RS5:
		return "1809"
	case osversion.V19H1:
		return "1903"
	case osversion.V19H2:
		return "1909"
	case osversion.V20H1:
		return "2004"
	case osversion.V20H2:
		return "2009"
	case osversion.V21H2Server:
		return "ltsc2022"
	default:
		panic("unsupported build")
	}
}

def getWindowsServerCoreImage(build uint16) string {
    tag := getContainerTagForContainerBuild(getContainerBuildForOSBuild(build))
    return "mcr.microsoft.com/windows/servercore:" + tag
}

def getWindowsNanoserverImage(build uint16) string {
    tag := getContainerTagForContainerBuild(getContainerBuildForOSBuild(build))
    return "mcr.microsoft.com/windows/nanoserver:"+ tag
}
```

And I'm assuming `getContainerTagForContainerBuild` changes when there's a new container image released, and `getContainerBuildForOSBuild` changes when there's a change in the rules about which host build can run which container builds.

(And also assuming we can simplify the current switch into "repository:tag", that might be an optimistic assumption)

Looking again, I realise this is just test code, so maybe I'm overthinking it. That said, doesn't this same logic exist elsewhere in the hcsshim tree already? Or maybe client code is expected to track the same mappings independently.

Reply:
The general idea is we try and include every new server related container image in this switch. Let's say you wanted to run this on a build that doesn't have a container image available (windows 11 insider build for example), then currently you can't as we'll just panic a couple lines in. This logic as least makes it so that if there's a container image that matches your host exactly, we'll pull that and if not and your build is higher than server 2022 where the abi should be stable, we can use the ws2022 image.

Reply:
Also never thought I'd see the day Python and Go become one 😝

Reply:
I fixed my `def` to be `func`. I'd been wondering why the syntax highlighting was not colouring that in...

Reply:
Let me explain my idea behind this change: When running our tests on preview builds (for which official container images aren't available yet) our tests panic. For any builds > V21H2 we can use the ltsc2022 image and continue tests instead of panicking. Notice, that when we are working with, say V25H2 preview builds, we will already have official container images for the build before that and those images would be used (via the default case) to run the tests (i.e we don't need to stick to V21H2 builds if V23H2 build is available).
So, from here onwards, every time an official container image is released, we will add a case for that in the switch and use that in our default "if build > V21H2" condition. Does that make sense?

Having said that, I like the idea of splitting the build tags from the repositories so that every time we add a new build, we have to update only 1 function instead of updating the two functions that we currently have. (However, I agree that this is just a test code and so I am fine even if we don't make this change right now).

Reply:
That makes sense to me, and is actually why I suggested this structure, to break the logic of "latest suitable container image version" out of "map build to tag". My thinking is that the `getContainerBuildForOSBuild` in the situation you've described is
```go
func getContainerBuildForOSBuild(build uint16) uint16 {
    if build >= osversion.V23H2Server {
        return osversion.V23H2Server
    } else if build >= osversion.V21H2Server {
        return osversion.V21H2Server
    }
    return build
}
```
so in-effect, unknown builds after the "downlevel support point" (osversion.V21H2Server) use the latest known image version, assuming that some time previously, `osversion.V23H2Server` was added to `getContainerTagForContainerBuild` when it became available.

And having put it like that, now I'm thinking what we could have (but it seems like over-engineering and at this point the conversation has probably consumed as much brain-power as all future updates to this code anyway...) would be a list of known container image builds, and you just search the list for the OS build you have, and take the highest-but-not-greater-than match. And then if the result was not an exact match, and is less than osversion.V21H2Server, `panic`.

And having put it like _that_, I realise that's roughly what the `getWindowsServercoreImage` would have evolved into, if we reversed it to be latest-first:
```go
func getWindowsServerCoreImage(build uint16) string {
	switch b := build; {
	case b >= osversion.V23H2Server: // Illustrative case for how this function grows over time
		return "mcr.microsoft.com/windows/servercore:ltsc2024"
	case b >= osversion.V21H2Server:
		return "mcr.microsoft.com/windows/servercore:ltsc2022"
	case b == osversion.V20H2:
		return "mcr.microsoft.com/windows/servercore:2009"
	case b == osversion.V20H1:
		return "mcr.microsoft.com/windows/servercore:2004"
	case b == osversion.V19H2:
		return "mcr.microsoft.com/windows/servercore:1909"
	case b == osversion.V19H1:
		return "mcr.microsoft.com/windows/servercore:1903"
	case b == osversion.RS5:
		return "mcr.microsoft.com/windows/servercore:1809"
	default:
		panic("unsupported build")
	}
}
```

----

Looking back, perhaps my faulty assumption is that after ltsc2024 images are created, we still want to support intermediate builds that are not ltsc2022, but use the ltsc2022 base image. That's the reason I see the `default` clause in the current code growing eternally, and want to promote it to be the primary behaviour, rather than being three nestings deep. If we're only going to keep the one `>` match, then I'm trying to solve the wrong problem. ^_^